deprecation fixes and a back port

raveit65 opened the pull request:
matekbd-indicator: don't use deprecated GtkAlignment
is if the applet use text for indicate the locale.
matekbd-indicator: don't use deprecated gdk_cairo_create
is when flags are displayed.
This can be easily tested with new mate-icon-theme-1.18.1 or 1.16.1 and you need to enable
```
[rave@mother ~]$ gsettings get org.mate.peripherals-keyboard-xkb.indicator show-flags
true
```

raveit65 commented:
The indicator applet is always installed but only visible if more than one keyboard layout is installed.
Install a second keyboard layout and you will see the applet in na-tray.
With clicking on it you can switch the layout.

monsta commented:
Yes, it's not related to indicator applet, it's actually a tray icon displayed by m-s-d automatically if you have more than one keyboard layout. Sometimes I think if it should be changed...

lukefromdc commented:
I'm not about to experiment with alternate keyboard layouts. Remember that all my experimentation is on my main encrypted system, and I type by touch. Don't want to risk a lockout or even having to shitcan and redo all the hidden files in /home

monsta commented:
I didn't pay much attention to that dialog, so it seems to me like it's the same as before, except for larger font on some keys:

![my-kbd](https://cloud.githubusercontent.com/assets/5138986/24697337/d911524a-19fd-11e7-8439-30f41d2b022b.png)

Old issues, like missing key borders or flickering on window resize, are still there, so I guess we can merge this without losing anything. :slightly_smiling_face:

monsta commented:
As for the flags, I can't get them to work. I have built and installed the latest mate-icon-theme, but still have issues.
First, changing icon theme via mate-appearance-properties does nothing to kbd indicator. (I needed to change the theme because my usual icon theme is Mint-X, not MATE or Menta.) So, after changing the icon theme, I need to set show-flags to false and back to true to make it actually load the new theme. I can also restart the session (to make m-s-d restart) so that the new icon theme would be loaded on the next session start.
Second, the actual loading of the flag fails for some reason... I only see the empty space there.

Tested with GTK+ 3.22.11.

raveit65 commented:
@monsta 
Thank for review, but non of this commits of the PR are related to keyboard-layout-view.
Fixes for the layout are already in master and you need to use latest mate-themes from git + editing mate-application.css, see https://github.com/mate-desktop/libmatekbd/pull/33#issue-218765513
Sadly, we need to fix missing key borders by themes.
I will update themes later for this.

raveit65 commented:
Only https://github.com/mate-desktop/libmatekbd/pull/34/commits/0d657a219549bda7704e9dcb5afcdce984b375b4
is related to the flags.
Here the flags are working with 1.16 without any changes from this PR  or previous changes in master.
1. Installing new releases from mate-icon-theme (1.16.1 /1.18.1)
2. updating icon-cache
3. installing a second keyboard layout
4. enable flags via gsettings key.

All this is working with libmate-keyboard-1.18.0/1.16.x tarballs on several installations in VMs or baremetal.

https://github.com/mate-desktop/libmatekbd/pull/34/commits/aca179c5e9464093aaacfe883eb4b337772acbaf

Is for using the indicator as text , ie. ru, de, en, etc.

https://github.com/mate-desktop/libmatekbd/pull/34/commits/0aa84c4aaf3196960c1fb2a5f8a39b34187ad16b
seems to be related to flags too, not sure.

Anyway, we can do a point release w/o this PR, i think Martin is waiting for it.
I thought the change from using expose_event to draw is important as it is a basic gtk2 to gtk3 port ?
So maybe you make only a code review for https://github.com/mate-desktop/libmatekbd/pull/34/commits/0d657a219549bda7704e9dcb5afcdce984b375b4  ?

monsta commented:
Text indicator looks the same as before, I guess no regressions happened.

monsta commented:
Icon caches are auto-updated on installing/upgrading .deb package of mate-icon-theme, so it was already done. I did it again manually just in case, no effect.

raveit65 commented:
@monsta 
Can you please check this dir?
/usr/share/icons/mate/scalable/animations
Do you see svg flags there?

monsta commented:
Yeah, lots of svg flags with 640x480 resolution. Why "animations" though? :slightly_smiling_face:

raveit65 commented:
Good, m-i-c is fine  :-)
I choose animations because this folder wasn't there and i had to use a folder from icon-theme specifications. Choosing 'flags' for example does not work here.
Can you please copy the flags to ~.icons/flags ?
Using this is defined in code, see https://github.com/mate-desktop/libmatekbd/blob/master/libmatekbd/matekbd-indicator-config.c#L233
Maybe there is a problem with using XDG icon theme folders in debian?
Or we need to define 
```
_build_filename (DATADIR,
			     "icons/<theme>/flags",
NULL));
```
or something else for debian.

raveit65 commented:
Or adding a specific dir for mate
"icons/mate-flags" ?
Where i can install the flags

raveit65 commented:
Ok, in ubuntu-17-04 with m-i-c.1.18.1 i get a warning about wrong icon format.
![keyboard-indicator-warning](https://cloud.githubusercontent.com/assets/961604/24703150/4e768fba-1a02-11e7-926e-27a89d00782d.png)
So it looks like the an application for reading svgs isn't installed out of box.
@monsta @flexiondotorg 
Which package is missing?
In fedora this works out of box.

monsta commented:
Oh, I get this error message too in Debian Testing here. Not always though. I found out it's best reproducible with resizing the panel...

monsta commented:
But I have librsvg installed, and eom can show these images.

monsta commented:
Actually, looks like the error message happens a lot more times that I noticed. It's just that the dialog is very often placed somewhere in the background, so I need to minimized all the apps to see it.

raveit65 commented:
I deleted the flags in /usr/share/........, moved them to ~.icons/flags and updated icon-cache, and i get the same warning about loading svgs with that folders.
So, it is a problem with loading svgs here.
You can try to use png flags from mintlocale, see this instructions https://github.com/mate-desktop/libmatekbd/issues/11#issuecomment-264155494
Of course you need to delete the svg flags and update icon-cache.
I will try this in ubuntu VM.

monsta commented:
Weird, I suddenly have flags working... :open_mouth: 
Only Russian though - US still gives an error. But eom opens both images just fine.

![ru-flag](https://cloud.githubusercontent.com/assets/5138986/24704149/6ab343a2-1a17-11e7-8521-4701a0d594ae.png)

raveit65 commented:
USA flags works here
![bildschirmfoto zu 2017-04-05 13-50-45](https://cloud.githubusercontent.com/assets/961604/24704233/0af10770-1a07-11e7-8450-af697afd6824.png)
and german
![bildschirmfoto zu 2017-04-05 13-53-00](https://cloud.githubusercontent.com/assets/961604/24704268/366ea8b2-1a07-11e7-854d-18885ce3bcc6.png)

...well this is really weird.

monsta commented:
I also built this w/o the last commit, and nothing changed. At least the problem isn't in this PR.

raveit65 commented:
Confirmed, russian federation flag loads out of box in ubuntu...
Is this issue related to this PR?
I don't think...

raveit65 commented:
Ok, than you agree with merging?

monsta commented:
For me it's ok, but maybe wait for more reviews?

raveit65 commented:
I am really in doubt that anyone else than you , luke or me will get noticed by this PR.
But Ok, i will tell Martin he can release 1.16.1 w/o this PR.
For fedora i can use expose_event --> draw fix as a patch.

willysr commented:
here's mine:


![libmatekbd](https://cloud.githubusercontent.com/assets/240100/24711461/3f839352-1a4a-11e7-9558-cad72ffea39e.jpg)

willysr commented:
ID works fine, but not US
![id](https://cloud.githubusercontent.com/assets/240100/24712125/2ce2f3c6-1a4c-11e7-9362-73aa1aa732b5.jpg)
![en-us](https://cloud.githubusercontent.com/assets/240100/24712124/2cda6d00-1a4c-11e7-86ee-fed9eba9208f.jpg)

raveit65 commented:
A us flag which is working in centos7
https://www.dropbox.com/s/bf5r2l369gc8wdv/us-new.svg?dl=0
I simply open ru flag and us flag in inskape and copied the content frm us flag in ru image , and renamed it to us flag.
So the culprit is definitely the image.
Open question is why this isn't a problem in fedora.
Any ways, that makes it not easier, than testing 250 flags and replacing some of them is a huge work.
Looking for another flag colection might be easier.
And i hat it to work with gimp or inkscape ;-)

raveit65 commented:
sigh, another prob with flags.
Some countries , ie. Argentinia, use 'spanish latin south-american' as layout.
And our gui selected the layout name and not the country name in this case   -_-

monsta commented:
Looks like we opened a can of worms :slightly_smiling_face:

raveit65 commented on libmatekbd/matekbd-indicator.c:
```diff
@@ -414,10 +410,7 @@ matekbd_indicator_prepare_drawing (MatekbdIndicator * gki, int group)
-		gtk_container_add (GTK_CONTAINER (align), label);
-		gtk_container_add (GTK_CONTAINER (ebox), align);
-
-		gtk_container_set_border_width (GTK_CONTAINER (align), 2);
```
Normal, i would set everything from the GtkAlignment to his parent, but the parent is a eventbox.
And in gtk-docs i found nothing to set a border on a eventbox.
Any idea?

cwendling replied:
[GtkEventBox](https://developer.gnome.org/gtk3/stable/GtkEventBox.html) is a [GtkContainer](https://developer.gnome.org/gtk3/stable/GtkContainer.html), so you can simply use the same API as with the alignment: `gtk_container_set_border_width (GTK_CONTAINER (ebox), 2);`.  Or alternatively you can use CSS stuff, or the [GtkWidget](https://developer.gnome.org/gtk3/stable/GtkWidget.html) API to set [margin](https://developer.gnome.org/gtk3/stable/GtkWidget.html#GtkWidget--margin).

raveit65 replied:
Opps, i was confused and said wrong things.
All settings from GtkAlignment are moved to GtkLabel.
Looks like GtkLabel isn't a GtkContainer, so i need to use GtkWidget API to set margin?

cwendling replied:
I guess setting the border on the event box which contains the label would have been the same, but margin is likely just fine too yeah.

raveit65 commented:
Ok, i used a border for GtkLabel, but i am not sure if the label is smaller now, please test.

For the flags i will create a branch for mate-icon-theme where we can replaced flags.
Help is welcome to create a list of broken flags and repair them like i said in https://github.com/mate-desktop/libmatekbd/pull/34#issuecomment-291940841
